### ISSUE TITLE
E: unexpected token tLABEL

### ISSUE THREAD

**Issue by envygeeks** [bbatsov/rubocop#2575]
Rubocop HEAD reports an error for valid Ruby:

```ruby
def tmpdir(*prefixes, root: nil)
  # Stuff
end
```

**Comment by envygeeks** [comment]
This turned out to be TargetRubyVersion: 1.9 as the default and I didn't notice that config.

**Comment by bbatsov** [comment]
It was introduced earlier today. @alexdowad I wonder if other people will be bitten by this, but I'm not sure how we can alleviate the confusion this change might cause.

**Comment by alexdowad** [comment]
This is tricky. If the inspected code was indeed supposed to run on Ruby 1.9, then it would be erroneous. We could try the newest Ruby parser as a fallback, and if it works, print a more informative message.

**Comment by envygeeks** [comment]
Parser does have CurrentRuby, which might be a route to take?

**Comment by alexdowad** [comment]
> Parser does have CurrentRuby, which might be a route to take?

No -- then we are back to being dependent on the interpreter which RC is running on. Then you can't run all the specs on all platforms. And you can't inspect any project if you are running RC on a lower version interpreter.

It would be better to fall back to the *newest* parser (currently `Parser::Ruby23`) and see if it can parse the code or not.

**Comment by envygeeks** [comment]
Let me translate what you just said (as I interpreted it) hoping that you will see the flaws: "I can't parse 2.3 from 2.0 (which `CurrentRuby` might bring) so I'll just go ahead and try and parse 2.3 from 2.0."

**Comment by envygeeks** [comment]
However, the reality of the situation is that `CurrentVersion` is probably the version of Ruby the user wants to target the most because it's the version of Ruby they have installed.

**Comment by alexdowad** [comment]
@envygeeks Relying on `CurrentVersion` is inherently flawed. I have no less than 12 versions of Ruby installed on my dev PC. If I happen to be using `2.2.2` in one shell, and `2.3.0` in another shell, I don't want RuboCop to give different results when I run it in *this* shell or *that* shell. I just want it to tell me whether my project code is good or not, according to my RC configuration. It doesn't matter what interpreter I happen to be running RC on.

Let's consider two situations:

Running RC with `TargetRubyVersion: 1.9`, source file:

```ruby
dafu32445u2  nqrjkfl nkl2-03294-p3
```

**Comment by envygeeks** [comment]
I'm abstaining.  I will work around this problem myself once you pick a design choice that's badly made.  That is all.  Thanks for you time.

**Comment by envygeeks** [comment]
I will ask at least one thing, throw out the default of 1.9.  It's dead, face it already.  It was EOL last year.

**Comment by alexdowad** [comment]
> I will ask at least one thing, throw out the default of 1.9. It's dead, face it already. It was EOL last year.

Good suggestion. I had actually set the default `TargetRubyVersion` to 2.0 at first, but changed it to 1.9 out of fear that we would break people's code. (Some projects still support 1.9, and if `TargetRubyVersion` is 2.0+ then some cops will auto-correct code to use 2.0+ syntax.)

Maintainers, what do you think?

If we say that RC will assume by default that all inspected projects are 2.0+, then it would be only natural for RC itself to drop support for 1.9.3 in the same release... are we ready for that?

**Comment by bbatsov** [comment]
Supporting 1.9.3 is not a big burden for us, that's why we haven't dropped the support yet. I'd like to see how many people are still clinging to it (e.g. some usage stats), before killing it.

As this release is pretty big, I'd rather drop the support in a subsequent release, so people on 1.9 can enjoy the boatload of improvements that are incoming.

**Comment by envygeeks** [comment]
For what it's worth, I would just like the default to be 2.0, whether 1.9 is removed has little impact on me as an end-user, but having 1.9 as the default has a huge impact on hundreds of projects I contribute to that will be impacted by this, all of which have started using at least basic keywords for some parts.  Especially once 1.9 went EOL.

**Comment by bbatsov** [comment]
Btw, here are some usage stats http://blog.scoutapp.com/articles/2015/12/29/state-of-the-2016-rails-stack

**Comment by alexdowad** [comment]
I agree that 2.0 makes much more sense as the default.

**Comment by jonas054** [comment]
Or even 2.1 based on those statistics. There should also be some kind of helpful message, "TargetRubyVersion is ... but the file ... needs TargetRubyVersion ... to be parseable" or something.

**Comment by bbatsov** [comment]
> Or even 2.1 based on those statistics. There should also be some kind of helpful message, "TargetRubyVersion is ... but the file ... needs TargetRubyVersion ... to be parseable" or something.

We'll have to reparse files on syntax errors to determine if something is parseable or not with another version of the parser, so I don't think this is worth it.

**Comment by alexdowad** [comment]
> I don't think this is worth it.

If you don't want it, that is fine, but I think it would be *much* better for the users.

**Comment by bbatsov** [comment]
My main concern is that at any syntax error we'll have to re-parse the file (potentially several times) to see if it will eventually become parseable. So this means RuboCop will likely get to some point where a file contains some unparseable code and spit the message then. And you have to decide if you'll display this message just once and exit and display it for every problematic file. 

It seems to me you'll get a better UX simply by printing a message in the beginning that RuboCop will interpret your code as Ruby 2.0 code, instead of introducing some complex logic.

**Comment by alexdowad** [comment]
Have a look at 9b96896.

**Comment by envygeeks** [comment]
Thanks!

**Comment by jonas054** [comment]
@alexdowad I think that will do fine.
